### ISSUE TITLE
Remove unneeded Android permissions

### ISSUE THREAD

**Issue by IzzySoft** [adapt-it/adapt-it-mobile#550]
On today's update, my scanner reported:

    ! repo/org.adaptit.adaptitmobile_49.apk declares sensitive permission(s): android.permission.READ_MEDIA_IMAGES android.permission.READ_MEDIA_AUDIO android.permission.READ_MEDIA_VIDEO android.permission.READ_EXTERNAL_STORAGE

![image](https://github.com/adapt-it/adapt-it-mobile/assets/6781438/65d050a3-7e45-43d6-b46d-184adc5c5e6b)

From the app description, I could understand `READ_EXTERNAL_STORAGE` to access `.txt` files. But what for is the media access needed? Thanks in advance for clarification!

Ps: The `DEPENDENCY_INFO_BLOCK` should be easy to get rid of if you use gradle:

```gradle
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains.

**Comment by eb1** [comment]
Hi Izzy -

Thanks for writing this issue up! 

Some background: Adapt It Mobile is an Apache Cordova app, which is essentially just a web app with hooks to native API calls where needed. Prior to the [1.14.0 release](https://github.com/adapt-it/adapt-it-mobile/releases/tag/v1.14.0), Adapt It Mobile scanned the local device for available files matching our supported file extensions (.txt, .sfm, .usfm, .xml, .aic, .lift, .tmx, .usx). We replaced that mechanism with a plugin that uses the native file chooser to access files -- much easier to work with, and doesn't feel weird.

I haven't investigated yet, but I suspect that those permissions were needed for the old mechanism and are no longer needed with the native dialogs. I'll update this issue once I find out more. Thanks!

**Comment by eb1** [comment]
Hi Izzy, I just verified that these permissions are no longer needed. I'll slate them for removal in the next update. Thanks again for letting me know!

**Comment by IzzySoft** [comment]
Nice! Thanks for the investigation and swift action! :star_struck: Will you take care for that BLOB as well? Not sure how that works with Cordova, though… How do you sign your APKs? That's where that BLOB is injected.

**Comment by eb1** [comment]
Let me see about the dependency block / blob stuff. Apache Cordova does use Gradle for Android, but it tends to lag behind the latest versions a bit AND it auto-injects into the build scripts. I don't see the dependency info as valid settings under Android Studio's Project Structure settings (https://stackoverflow.com/questions/77815830/what-does-retain-information-about-dependencies-in-the-apk-bundle-do-in-androi). It's possible I'll need to manually add the settings in.

**Comment by IzzySoft** [comment]
> I don't see the dependency info as valid settings under Android Studio's Project Structure settings

No, you won't find it there. You'd need to add it to your `build.gradle`. An alternative would be using `apksigner` to sign the APK.

**Comment by eb1** [comment]
@IzzySoft - I've split off the issue to #552 . In talking with the Cordova devs, it sounds like [they're using an included `.gradle` file](https://cordova.apache.org/docs/en/12.x/guide/platforms/android/index.html#extending-buildgradle) that doesn't get overwritten with each build -- this is where I need to place the `dependenciesInfo` block. 

A question for you: do you happen to know of a tool to verify that the block has been removed? Just want to ensure that the dependency blob actually is gone. Thanks!

**Comment by IzzySoft** [comment]
> do you happen to know of a tool to verify that the block has been removed?

Sure, how else should I have figured it? And as my code is FOSS, too, you can find it in my GitLab repo: [`CheckSigningBlocks.py`](https://gitlab.com/IzzyOnDroid/repo/-/blob/master/lib/CheckSigningBlocks.py) (next to several other checker scripts :wink:) Run with `-h` for help/usage-hints – or just look at the end of the file for them.

**Comment by IzzySoft** [comment]
Today the scanner just complains about a single missing explanation:

    ! repo/org.adaptit.adaptitmobile_50.apk declares sensitive permission(s): android.permission.READ_EXTERNAL_STORAGE

What shall we do with that?

**Comment by IzzySoft** [comment]
> track the fix in the next patch

Wonderful, thanks!

> weirdness with the listing for Adapt It Mobile on the IzzyOnDroid

Suggestion: I open a PR providing the metadata set up with my repo. It's using Fastlane structures, which are pretty much standard for that. Then after merging that, you can adjust everything as you see it fit (be welcome to use my [Fastlane Cheat Sheet](https://gitlab.com/-/snippets/1895688) for guidance with that). When ready let me know, and I enable it on my end – which means for each release after that, your Fastlane would be pulled along.

> Where is that anti-feature message coming from?

That must have been pointed out somewhere at the time the app was added to IoD. Which is more than a year ago, so I don't remember all the details. Probably deduced from the linked page of the "original project".

> There's no such thing as an "Adapt server" -- all the data is held on the mobile device in a SQLite database.

Then I'll remove that AF now. Effective with the next sync at 7 pm UTC. Thanks for pointing out! Might be a good idea having that in the app description then as well :wink:

**Comment by IzzySoft** [comment]
Looks like you have added them, @eb1? Today the scanner reported:

```
! repo/org.adaptit.adaptitmobile_57.apk declares sensitive permission(s):
  android.permission.READ_EXTERNAL_STORAGE
  android.permission.READ_MEDIA_IMAGES
  android.permission.READ_MEDIA_VIDEO
  android.permission.READ_MEDIA_AUDIO
```

What is their purpose, so I can check to add them to your app's "green list"?

**Comment by IzzySoft** [comment]
Either the app description is not up-to-date, or I must miss something (or both, who knows):

![image](https://github.com/user-attachments/assets/be3dae78-c1df-413e-a205-4017201d0503)

No word of images, videos or audio files there. So the app should not even announce a capability to receive such files, as what would it do with them? Guess I'm a bit confused now – so I really appreciate details on this :wink:

**Comment by eb1** [comment]
The app description is correct. Adapt It Mobile does only support / read text based content. We don't currently support images, videos, or audio -- and do not show up in the sharing panel as an app that would accept those types of data.

The issue is with a 3rd party plugin that treats the permissions as monolithic -- either you have asked for all the permissions (in which case it will convert the path from `content://` to `file://`), or it will pass back a permissions exception.

This afternoon I did a fork of the plugin (https://github.com/VIAVI-Solutions/cordova-plugin-filepath), to see if the permission request is _really_ needed in order to perform the conversion from `content://` to `file://`. If it isn't needed, I might be able to publish and point to the fork and remove the permissions via a patch to Adapt It Mobile. No guarantees, though!

**Comment by IzzySoft** [comment]
No guarantees expected – but your help and efforts highly appreciated, thanks!

**Comment by eb1** [comment]
Update on this issue:

I was able to fork the 3rd party plugin and rework it so that it only requested permissions for files of mime type `video/*`, `audio/*`, and `image/*`. I tested the changes on API 29 and 35, both for text and image content; it did appear to work properly in all permutations I tested. More importantly, **Adapt It Mobile was able to open up shared text content under API 29 and 34 without requiring additional permissions.**

I've created PR https://github.com/VIAVI-Solutions/cordova-plugin-filepath/pull/1 against the plugin, but I do have our forked repo to point to if there's no response. The next patch should have the change.
